src/Control/SessionHandler/FileSessionHandler.php

```diff
@@ -0,0 +1,351 @@
+     */
+    private function setSavePath(string $path): void
+    {
+        // Handle configuration for depth and mode arguments
+        // See https://www.php.net/manual/en/session.configuration.php#ini.session.save-path for details
+        if (str_contains($path, ';')) {
+            $parts = explode(';', $path);
+            $numParts = count($parts);
+            if ($numParts > 3) {
+                throw new InvalidArgumentException('$path was invalid');
+            }
+
+            // Set n depth
+            $this->numSubDirs = (int) array_shift($parts);
+
+            // Mode is the 2nd item in the config. Usually will be like '600' but we need true octal e.g. 0600
+            if ($numParts === 3) {
+                $this->mode = $this->ensureValueIsOctal(array_shift($parts));
+            }
+
+            // The actual path is the last part of the configuration.
+            $path = array_shift($parts);
+        }
+        $this->baseDir = $path;
+    }
+
+    /**
+     * Get the absolute path to the file which this session's data will be stored in.
+     */
+    private function getFilePath(#[SensitiveParameter] string $id): string
+    {
+        $subParts = str_split(substr($id, 0, $this->numSubDirs));
+        return Path::join(...[$this->baseDir, ...$subParts, FileSessionHandler::SESSION_FILE_PREFIX . $id]);
+    }
+
+    /**
+     * Take a numeric string from save_path configuration and make sure it's a PHP octal integer.
+     */
+    private function ensureValueIsOctal(string $value): int
+    {
+        if (strlen($value) < 4) {
+            $value = '0' . $value;
+        }
+        return intval($value, 8);
+    }
+
+    /**
+     * Check if a session file needs its permissions to be updated to match configured mode.
+     */
+    private function needsPermissionUpdate(#[SensitiveParameter] string $path): bool
+    {
+        // Silence PHP's warning, any problems here will already be logged by the calling method.
+        $rawPerms = @fileperms($path);
+        // If we had trouble reading the permissions, just assume it needs updating.
+        // Either we'll be able to set it right, or the calling method will fail and log.
+        if ($rawPerms === false) {
+            return true;
+        }
+        // See https://www.php.net/manual/en/function.fileperms.php#refsect1-function.fileperms-examples
+        $permsInOctal = intval(substr(sprintf('%o', $rawPerms), -4), 8);
+        return $permsInOctal !== $this->mode;
+    }
+
+    /**
+     * Get the exception message without leaking session IDs
+     */
+    private function getSafeExceptionMessage(Exception $exception, #[SensitiveParameter] ?string $id = null): string
+    {
+        if ($id) {
+            return str_replace($id, '<session_id>', $exception->getMessage());
+        }
+        $prefix = FileSessionHandler::SESSION_FILE_PREFIX;
+        return preg_replace('/(' . $prefix . ')[\w_-]+/', '$1<session_id>', $exception->getMessage());
```

Comment:
```suggestion
        return preg_replace('/(' . $prefix . ')[\w\-]+/', '$1<session_id>', $exception->getMessage());
```
_ is part of \w
escaping the - as \- technically isn't needed though it's a little less confusing IO

Reply:
Also are we 100% sure that `\w\-` will always cover the session ID? any chance of a utf character ever getting in there?

Should we be validating $id is `\w\-` in every method that accepts $id as a param and throwing/logging an exception if not?

Reply:
> are we 100% sure that \w\- will always cover the session ID? any chance of a utf character ever getting in there?

The session ID is generated based on the `session.sid-bits-per-character` ini configuration which specifies the valid characters: https://www.php.net/manual/en/session.configuration.php#ini.session.sid-bits-per-character

> Should we be validating $id is \w\- in every method that accepts $id as a param and throwing/logging an exception if not?

I don't think that's necessary.
All valid character sets are caught with `\w\-` other than I guess if you set it to `6` which might include `,` which I have now added to the regex.